### ISSUE TITLE
Add annotations/tags/baggage support to Spans

### ISSUE THREAD

**Issue by nicmunroe** [Nike-Inc/wingtips#23]
The Dapper paper and many distributed tracing impls have support for arbitrary annotations/tags/baggage. Wingtips should follow suit to give users more flexibility in what wingtips does for them, compatibility with other impls, etc.

This may tie in with OpenTracing support.

**Comment by adriancole** [comment]
One note is that arbitrary baggage isnt always a part of tracing apis. For
example in baggage itself (the work from brown) tracing is a consumer of
baggage propagation layer. This is the same in finagle and grpc, for
example, who have their own context propagation layer independent of
tracing.

In other words, tags/annotatioms (dimensional lookup keys) are far more
common than baggage in a tracing api.

That said, definitely worth considering as it is convient for folks to be
able to stuff things like tokens into a distributed context and currently
the OT api defines a hook for that.

**Comment by nicmunroe** [comment]
Yeah this ticket is more of a placeholder for me to go investigate the current landscape and decide what I want to implement in wingtips for the near term. It will probably be along the simpler annotations line than full-throated baggage support. At least initially. Just a guess until I can carve out some time to investigate though.

**Comment by nicmunroe** [comment]
Tags are now supported after #74 and #78. Annotations and updating instrumentation to automatically tag/annotate (i.e. automatically tagging spans with `http.method`, `http.path`, `http.status_code`, `error`, etc) will come later.

**Comment by nicmunroe** [comment]
Annotations added in #84.

Wingtips version `0.16.0` added general tag support, `0.17.0` added auto-tagging for various server/client Wingtips instrumentations, and `0.18.0` added general annotation support.

Baggage has a lot of complexity and difficulty associated with it, and the benefit (and demand) doesn't seem worth it at this time. So for the purpose of this issue, we'll call it good after tags and annotations. If baggage needs to be revisited, it will be done in a separate issue.
